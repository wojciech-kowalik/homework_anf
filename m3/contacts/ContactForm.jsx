/* eslint-disable */
import React from 'react';

export class ContactForm extends React.Component {

  static defaultProps = {
    contact: {
      name: '',
      details: ''
    }
  }

  onChange = (e) => {
    let target = e.target;
    let name = target.name;
    let value = target.type === "checkbox" ? target.checked : target.value;

    this.props.onChange({
      ...this.props.contact,
      [name]: value
    });
  }

  onSubmit = (e) => {
    e.preventDefault();
    this.props.onSubmit(this.props.contact);
  }

  render() {
    return (
      <form onSubmit={this.onSubmit}>
        <div className="form-group">
          <label className="d-block w-100">
            Name:
            <input
              className="form-control"
              name="name"
              value={this.props.contact.name}
              onChange={this.onChange}
            />
          </label>
        </div>
        <div className="form-group">
          <label className="d-block w-100">
            Contact details:
            <textarea
              className="form-control"
              name="details"
              value={this.props.contact.details}
              onChange={this.onChange}
            />
          </label>
        </div>
        <div className="form-group">
          <input
            type="button"
            id="cancel-contact"
            className="btn btn-default"
            value="Cancel"
            onClick={this.props.onCancel}
          />
          <input
            type="submit"
            id="save-contact"
            className="btn btn-primary ml-2"
            value="Save"
          />
        </div>
      </form>
    );
  }
}
