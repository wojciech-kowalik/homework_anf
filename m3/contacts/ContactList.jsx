/* eslint-disable */
import React from 'react';

export class ContactList extends React.Component{

  static defaultProps = {
    contacts: []
  }

  render() {
    return (
      <div className="list-group">
        {this.props.contacts.map((contact) => (
          <button
            key={contact.id}
            onClick={(e) => this.props.onSelect(contact)}
            className={`list-group-item ${this.props.selected && (this.props.selected.id === contact.id) ? 'active' : ''}`}
          >
            {contact.name}
          </button>
        ))}
      </div>
    );
  }
}
