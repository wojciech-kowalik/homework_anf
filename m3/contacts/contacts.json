[
	{
		"id": "ad324356-384a-4fa4-a1d9-e373b0f760d0",
		"name": "Allison Clayton",
		"details": "383-897-142"
	},
	{
		"id": "cbe348af-a5ae-47c2-8d2b-07441a773fb7",
		"name": "David Jacobs",
		"details": "david.jacobs@erntogra.com"
	},
	{
		"id": "4822e046-eee3-4d3a-b9be-f860f48ea5df",
		"name": "Dominic Ramsey",
		"details": "117 Aster Court APT 187\nMiami 78499"
	},
	{
		"id": "3d3a4d97-0bad-4ea7-93e0-e936dcc02928",
		"name": "Victoria Clarke",
		"details": "victoria.clarke@buapel.com"
	},
	{
		"id": "4bd7bd06-c5c5-428f-bde9-de07e59628e7",
		"name": "Zoe M. Russell",
		"details": "zoe.russell@furba.biz\n014-203-3635"
	}
]
