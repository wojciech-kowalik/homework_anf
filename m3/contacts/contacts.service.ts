import { v4 as uuid } from "uuid";

export interface Contact {
  id: string;
  name: string;
  details: string;
}

export interface ServiceInterface<T> {
  getAll: () => Promise<T[]>;
  update: (updateCallback: (data: T[]) => void) => void;
  save: (entity: T) => Promise<T>;
}

export class ContactsService implements ServiceInterface<Contact> {
  constructor(private contacts: Contact[]) {}

  async getAll() {
    return this.contacts;
  }

  update(updateCallback: (contacts: Contact[]) => void) {
    const clone = [...this.contacts];
    updateCallback(clone);
    this.contacts = clone;
  }

  async save(contact: Contact) {
    const index = this.contacts.findIndex(({ id }) => contact.id === id);

    if (index === -1) {
      contact.id = uuid();
      this.update((contacts) => contacts.push(contact));
    } else {
      this.update((contacts) => contacts.splice(index, 1, contact));
    }

    return contact;
  }
}
