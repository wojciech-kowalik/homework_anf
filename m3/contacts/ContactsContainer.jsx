/* eslint-disable */
import React from 'react';

import { ContactList } from './ContactList';
import { ContactForm } from './ContactForm';

export class ContactsContainer extends React.Component {

  constructor(props) {
    super(props);

    // ContactsService object
    this.service = this.props.service;

    this.state = {
      contacts: [],
      selected: null
    };

    this.service.getAll().then((contacts) => {
      this.setState({
        contacts
      });
    });
  }

  newContact = () => {
    this.setState({
      selected: {
        name: '',
        details: ''
      }
    });
  }

  onSelect = (contact) => {
    this.setState({
      selected: contact
    });
  }

  onSubmit = (contact) => {
    this.service.save(contact).then(()=>{
      this.service.getAll().then((contacts) => {
        this.setState({
          contacts
        });
      });
    });
  }

  onCancel = () => {
    this.setState({
      selected: null
    });
  }

  render() {
    return (
      <div className="container">
        <div className="row">
          <div className="col-md-12">
            <h1>Contacts</h1>
          </div>
        </div>
        <div className="row">
          <div className="col-md-4">
            <ContactList
              contacts={this.state.contacts}
              selected={this.state.selected}
              onSelect={this.onSelect}
            />
          </div>
          <div className="col-md-4">
            {this.state.selected
            ? (
              <ContactForm contact={this.state.selected}
                onChange={this.onSelect}
                onSubmit={this.onSubmit}
                onCancel={this.onCancel}
              />
            ) : (
              <div>
                <button id="new-contact" onClick={this.newContact} className="btn btn-primary">New contact</button>
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}
