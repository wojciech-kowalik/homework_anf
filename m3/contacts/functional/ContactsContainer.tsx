import React, { useEffect, useState } from "react";

import { ContactList } from "../functional/ContactList";
import { ContactForm } from "../functional/ContactForm";
import { ServiceInterface, Contact } from "../contacts.service";

interface ContactsContainerProps {
  service: ServiceInterface<Contact>;
}

export const ContactsContainer: React.FC<ContactsContainerProps> = (
  props
): JSX.Element => {
  const { service } = props;
  const [contacts, setContacts] = useState<Contact[]>([]);
  const [selected, setSelected] = useState<Contact | null>(null);
  const defaultValue: Contact = { id: "", name: "", details: "" };

  useEffect(() => {
    service.getAll().then((data) => {
      setContacts(data);
    });
  }, [service]);

  const newContact = () => {
    setSelected(defaultValue);
  };

  const onSelect = (contact: Contact) => {
    setSelected(contact);
  };

  const onSubmit = (contact: Contact) => {
    service.save(contact).then(() => {
      service.getAll().then((data) => {
        setContacts(data);
      });
    });
  };

  const onCancel = () => {
    setSelected(null);
  };

  return (
    <div className="container">
      <div className="row">
        <div className="col-md-12">
          <h1>Contacts</h1>
        </div>
      </div>
      <div className="row">
        <div className="col-md-4">
          <ContactList
            contacts={contacts}
            selected={selected}
            onSelect={onSelect}
          />
        </div>
        <div className="col-md-4">
          {selected ? (
            <ContactForm
              contact={selected}
              onChange={onSelect}
              onSubmit={onSubmit}
              onCancel={onCancel}
            />
          ) : (
            <div>
              <button
                id="new-contact"
                onClick={newContact}
                className="btn btn-primary"
              >
                New contact
              </button>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};
