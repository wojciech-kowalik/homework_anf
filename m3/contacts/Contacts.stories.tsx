/* eslint-disable import/first */
import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Meta } from '@storybook/react/types-6-0';

import contacts from './contacts.json';
import { ContactsContainer as FunctionalContainer } from './functional/ContactsContainer';
import { ContactsContainer } from "./ContactsContainer";
import { ContactsService } from './contacts.service';

const svc = new ContactsService(contacts);

import { lessons } from 'stories';
export default {
  title: lessons.m3.add('HOMEWORK').toString(),
  argTypes: {
  },
} as Meta;

export const Contacts = () => {
  return (
    <>
      <link
        rel="stylesheet"
        href="https://maxcdn.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css"
        integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO"
        crossOrigin="anonymous"
      />
      <ContactsContainer service={svc} />
      <FunctionalContainer service={svc} />
    </>
  );
}
