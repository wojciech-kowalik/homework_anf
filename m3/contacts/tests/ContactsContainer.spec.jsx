/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
import React from 'react';
import flushPromises from 'flush-promises';
import { act } from "react-dom/test-utils";
import { mount } from 'enzyme';
import { ContactsContainer } from '../ContactsContainer';
import { ContactsService } from '../contacts.service';

describe('ContactsContainer Component', () => {
  function createMockService() {
    const data = [{
      id: '4567', name: 'candidate test', details: 'candidate test'
    }];

    const svc = new ContactsService(data);
    jest.spyOn(svc, 'getAll');
    jest.spyOn(svc, 'save');
    return svc;
  }

  it('should match snapshot', () => {
    const mockService = createMockService();
    const wrapper = mount(<ContactsContainer service={mockService} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call contactsService.getContacts() when rendered', () => {
    const mockService = createMockService();
    const wrapper = mount(<ContactsContainer service={mockService} />);

    expect(mockService.getContacts.mock.calls.length).toEqual(1);

    return mockService.getContacts.mock.results[0].value.then(() => {
      wrapper.update();
      expect(wrapper.find('.list-group-item').length).toEqual(mockService.contacts.length);
    });
  });

  it('should call contactsService.saveContact() when form is submitted', () => {
    const mockService = createMockService();
    const wrapper = mount(<ContactsContainer service={mockService} />);
    expect(mockService.getContacts.mock.calls.length).toEqual(1);

    wrapper.setState({ selected: mockService.contacts[0] });
    wrapper.find('form [type="submit"]').simulate('click');
    wrapper.find('form').simulate('submit');
    expect(mockService.saveContact.mock.calls[0]).toEqual([mockService.contacts[0]]);
  });

  it('should fetch contacts from contactsService and update contacts list when form is submitted', () => {
    const mockService = createMockService();
    const wrapper = mount(<ContactsContainer service={mockService} />);
    expect(mockService.getContacts.mock.calls.length).toEqual(1);

    return mockService.getContacts.mock.results[0].value.then(() => {
      wrapper.update();
      expect(wrapper.find('.list-group-item').length).toEqual(mockService.contacts.length);
    });
  });

  it('should add new contact to list when new contact is saved', () => {
    const mockService = createMockService();
    const wrapper = mount(<ContactsContainer service={mockService} />);

    wrapper.setState({ selected: { name: 'new', details: 'new' } });
    wrapper.find('form [type="submit"]').simulate('click');
    wrapper.find('form').simulate('submit');

    expect(mockService.saveContact.mock.calls.length).toEqual(1);

    // Wait for saveContact
    return Promise.resolve(mockService.saveContact.mock.results[0].value)
      .then(() => {
        expect(mockService.getContacts.mock.calls.length).toEqual(2);

        // Wait for second getContacts
        return mockService.getContacts.mock.results[1].value;
      })
      .then(() => {
        wrapper.update();
        expect(wrapper.find('.list-group-item').length).toEqual(mockService.contacts.length);
      });
  });

  it('should update contact on the list when existing contact is saved', () => {
    const mockService = createMockService();
    const wrapper = mount(<ContactsContainer service={mockService} />);
    const contact = mockService.contacts[0];
    contact.name = 'changed name';

    wrapper.setState({ selected: contact });
    wrapper.find('[name="name"]').simulate('change');

    wrapper.find('form [type="submit"]').simulate('click');
    wrapper.find('form').simulate('submit');

    expect(mockService.saveContact.mock.calls.length).toEqual(1);

    // Wait for saveContact
    return Promise.resolve(mockService.saveContact.mock.results[0].value)
      .then(() => {
        expect(mockService.getContacts.mock.calls.length).toEqual(2);

        // Wait for second getContacts
        return mockService.getContacts.mock.results[1].value;
      })
      .then(() => {
        wrapper.update();
        expect(wrapper.find('.list-group-item').at(0).text()).toBe(contact.name);
      });
  });
});
