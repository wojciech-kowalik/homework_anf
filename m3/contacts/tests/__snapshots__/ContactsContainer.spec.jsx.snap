// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsContainer Component should match snapshot 1`] = `
<ContactsContainer
  service={
    ContactsService {
      "contacts": Array [
        Object {
          "details": "candidate test",
          "id": "4567",
          "name": "candidate test",
        },
      ],
      "getContacts": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "saveContact": [MockFunction],
    }
  }
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <h1>
          Contacts
        </h1>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-4"
      >
        <ContactList
          contacts={Array []}
          onSelect={[Function]}
          selected={null}
        >
          <div
            className="list-group"
          />
        </ContactList>
      </div>
      <div
        className="col-md-4"
      >
        <div>
          <button
            className="btn btn-primary"
            id="new-contact"
            onClick={[Function]}
          >
            New contact
          </button>
        </div>
      </div>
    </div>
  </div>
</ContactsContainer>
`;
